data_safe_haven/external/interface/azure_postgresql_database.py

```diff
@@ -198,28 +192,36 @@ def set_database_access(self, action: str) -> None:
             )
         elif action == "disabled":
             self.logger.debug(
-                f"Removing temporary firewall rule for [green]{self.current_ip}[/]...",
-            )
-            self.wait(
-                self.db_client.firewall_rules.begin_delete(
-                    self.resource_group_name, self.server_name, rule_name
-                )
+                f"Removing all firewall rule(s) from [green]{self.server_name}[/]...",
             )
-            self.wait(
-                self.db_client.servers.begin_update(
-                    self.resource_group_name,
-                    self.server_name,
-                    ServerUpdateParameters(public_network_access="Disabled"),
-                )
+            rules = cast(
+                list[FirewallRule],
+                self.db_client.firewall_rules.list_by_server(
+                    self.resource_group_name, self.server_name
+                ),
             )
+            for rule in rules:
+                if rule.name:
```

Comment:
Is it possible for a rule to have no name attribute?
If it doesn't do we really want to just skip it?

Reply:
I think it *is* possible (but unlikely) for a rule to have no name. If this happens we can't delete it as `begin_delete` takes a rule name as input.

An equivalent workaround would be to wrap the `begin_delete` in a `try/except` - would you prefer this?

Reply:
Might be better to give a warning rather than raise an exception?

I have made some changes around what I was thinking in 0704ba2. What do you think?

Reply:
Looks good. I've added an explicit type-cast to `str` as my type-checker insists that `rule.name` is always `None` since it's initialised as `None` and no functions explicitly change it (`az` libraries seem to do a lot of direct editing of the `__dict__` of objects).